File: frontend/lib/models/tasdiqlash_tarixi.dart
import 'tasdiq_yozuvi.dart';

/// `GET /tasdiqlash-tarixi` javobi — kamera-tasdiq so'rovlari va
/// kip-to'g'rilash zayavkalarini BITTA normallashtirilgan qatorga
/// birlashtirgan yozuv. `tur` ('kamera' | 'kip_togrilash') qaysi asl
/// jadvaldan kelganini bildiradi; shu maydon orqali [amalEndpointi] ham
/// tegishli backend endpointini (tasdiqlash/rad-etish uchun) tanlaydi —
/// shunda IKKITA turli model o'rniga bitta model bilan
/// `TasdiqRoyxatiEkrani<TasdiqTarixiYozuvi>`ni ishlatish mumkin.
class TasdiqTarixiYozuvi implements TasdiqYozuvi {
  @override
  final int id;
  final String tur;
  final DateTime vaqt;
  final String operatorIsm;
  final String tavsif;
  final String? sabab;
  @override
  final String holati; // kutilmoqda | tasdiqlangan | rad_etilgan
  final DateTime? halQilinganVaqt;
  @override
  final String? halQilganIsm;
  @override
  final String? halQilishManbasi;
  final String? izoh;
  final bool dublikatShubhasi;

  TasdiqTarixiYozuvi({
    required this.id,
    required this.tur,
    required this.vaqt,
    required this.operatorIsm,
    required this.tavsif,
    this.sabab,
    required this.holati,
    this.halQilinganVaqt,
    this.halQilganIsm,
    this.halQilishManbasi,
    this.izoh,
    this.dublikatShubhasi = false,
  });

  @override
  bool get kutilmoqda => holati == 'kutilmoqda';

  bool get kameraTuri => tur == 'kamera';

  String get amalEndpointi => kameraTuri ? '/kamera-tasdiq' : '/kip-togrilash';

  factory TasdiqTarixiYozuvi.fromJson(Map<String, dynamic> j) => TasdiqTarixiYozuvi(
        id: j['id'],
        tur: j['tur'],
        vaqt: DateTime.parse(j['vaqt']),
        operatorIsm: j['operator_ism'],
        tavsif: j['tavsif'],
        sabab: j['sabab'],
        holati: j['holati'],
        halQilinganVaqt: j['hal_qilingan_vaqt'] == null ? null : DateTime.parse(j['hal_qilingan_vaqt']),
        halQilganIsm: j['hal_qilgan_ism'],
        halQilishManbasi: j['hal_qilish_manbasi'],
        izoh: j['izoh'],
        dublikatShubhasi: j['dublikat_shubhasi'] == true,
      );
}

File: backend/app/schemas/tasdiqlash_tarixi.py
import enum
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class TasdiqTarixiHolati(str, enum.Enum):
    """`KameraTasdiqHolati` va `KipTogrilashHolati`ning QIYMATLARI bir xil
    (kutilmoqda/tasdiqlangan/rad_etilgan) — shu umumiy filtr uchun mustaqil,
    ikkalasidan bir xil qiymat bilan qurilishi mumkin bo'lgan enum."""

    kutilmoqda = "kutilmoqda"
    tasdiqlangan = "tasdiqlangan"
    rad_etilgan = "rad_etilgan"


class TasdiqTarixiYozuvi(BaseModel):
    """`GET /tasdiqlash-tarixi` — kamera-tasdiq so'rovlari va kip-to'g'rilash
    zayavkalarini BITTA normallashtirilgan qatorga birlashtirgan javob.
    `tur` qaysi asl jadvaldan kelganini bildiradi — frontend amallarni
    (tasdiqlash/rad-etish) shunga qarab tegishli `/kamera-tasdiq` yoki
    `/kip-togrilash` endpointiga yo'naltiradi."""

    tur: Literal["kamera", "kip_togrilash"]
    id: int
    vaqt: datetime
    operator_ism: str
    tavsif: str  # nima haqida: mahsulot/partiya (+kamera uchun og'irlik) yoki eski->yangi
    sabab: str | None  # faqat kip_togrilash'da (operator ko'rsatgan sabab)
    holati: TasdiqTarixiHolati
    hal_qilingan_vaqt: datetime | None
    hal_qilgan_ism: str | None
    hal_qilish_manbasi: str | None
    izoh: str | None  # hal qiluvchining (admin) izohi, faqat rad etishda ixtiyoriy kiritiladi
    # Faqat tur="kamera" va hali "kutilmoqda" qatorlar uchun hisoblanadi (qarang
    # GET /kamera-tasdiq'dagi bir xil naqsh) — kip_togrilash uchun har doim False.
    dublikat_shubhasi: bool = False
